File: database_integration_scripts/enhance_db/src/bin/45_create_consolidate.rs
/// Create a `consolidate` table that brings together key data for each individual
/// who belongs to at least one Cliopatria polity.
///
/// Columns:
///   wikidata_id, name_en, impact_year, polity_name, occupations (;-separated),
///   gender, references_count, is_scientist (0/1), is_artist (0/1)
///
/// Phase 1: Build occupation_id -> meta_occupation lookup
/// Phase 2: Bulk INSERT from individuals_cliopatria JOIN individuals
/// Phase 3: Update is_scientist / is_artist from individuals_keys + occupations lookup
/// Phase 4: Create indexes
/// Phase 5: Export to CSV
/// Phase 6: Statistics
use anyhow::Result;
use indicatif::{ProgressBar, ProgressStyle};
use rusqlite::{params, Connection};
use std::collections::HashMap;
use std::fs;
use std::io::Write;
use std::time::Instant;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const CSV_PATH: &str = "data/consolidate.csv";
const TASK_LOG: &str = "task.log";
const BATCH_SIZE: usize = 50_000;

fn log(msg: &str) {
    println!("{}", msg);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", msg).unwrap();
}

fn elapsed(start: Instant) -> String {
    let s = start.elapsed().as_secs();
    if s < 60 {
        format!("{}s", s)
    } else if s < 3600 {
        format!("{}m {}s", s / 60, s % 60)
    } else {
        format!("{}h {}m {}s", s / 3600, (s % 3600) / 60, s % 60)
    }
}

/// Escape a field for CSV output (RFC 4180)
fn csv_escape(field: &str) -> String {
    if field.contains(',') || field.contains('"') || field.contains('\n') || field.contains('\r') {
        format!("\"{}\"", field.replace('"', "\"\""))
    } else {
        field.to_string()
    }
}

fn main() -> Result<()> {
    let _ = fs::remove_file(TASK_LOG);
    let total_start = Instant::now();
    log("=== Step 45: Create consolidate table ===");

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    // ========================================================
    // PHASE 1: Build occupation_id -> meta_occupation lookup
    // ========================================================
    log("[45] Phase 1: Building occupation lookup...");
    let mut occ_lookup: HashMap<String, String> = HashMap::new();
    {
        let mut stmt = conn.prepare(
            "SELECT id, meta_occupation FROM occupations WHERE meta_occupation IS NOT NULL",
        )?;
        let rows = stmt.query_map([], |r| {
            Ok((r.get::<_, String>(0)?, r.get::<_, String>(1)?))
        })?;
        for r in rows {
            let (id, meta) = r?;
            occ_lookup.insert(id, meta);
        }
    }
    log(&format!(
        "[45] Occupation lookup: {} entries",
        occ_lookup.len()
    ));

    // ========================================================
    // PHASE 2: Create and populate consolidate table (bulk INSERT)
    // ========================================================
    log("[45] Phase 2: Creating and populating consolidate table...");
    let phase2_start = Instant::now();

    conn.execute_batch("DROP TABLE IF EXISTS consolidate;")?;
    conn.execute_batch(
        "CREATE TABLE consolidate (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            impact_year INTEGER,
            polity_name TEXT,
            occupations TEXT,
            gender TEXT,
            references_count INTEGER,
            is_scientist INTEGER DEFAULT 0,
            is_artist INTEGER DEFAULT 0
        );",
    )?;
    log("[45] Table created, now bulk inserting...");

    // Bulk insert using JOIN — only individuals that belong to at least one polity
    let inserted = conn.execute(
        "INSERT INTO consolidate (wikidata_id, name_en, impact_year, polity_name, occupations, gender, references_count)
         SELECT
             ic.wikidata_id,
             ic.name_en,
             ic.impact_date,
             ic.polity_name,
             i.occupations_en,
             i.gender,
             i.identifiers_count
         FROM individuals_cliopatria ic
         JOIN individuals i ON ic.wikidata_id = i.wikidata_id",
        [],
    )?;
    log(&format!(
        "[45] Phase 2: Inserted {} rows ({})",
        inserted,
        elapsed(phase2_start)
    ));

    // ========================================================
    // PHASE 3: Update is_scientist / is_artist flags
    // ========================================================
    let total: i64 = conn.query_row("SELECT COUNT(*) FROM consolidate", [], |r| r.get(0))?;
    log(&format!(
        "[45] Phase 3: Updating is_scientist/is_artist for {} rows...",
        total
    ));

    let pb = ProgressBar::new(total as u64);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
            .unwrap(),
    );
    pb.set_message("Updating scientist/artist flags");

    let mut offset: i64 = 0;
    let mut scientist_count = 0u64;
    let mut artist_count = 0u64;
    let mut updated_count = 0u64;

    loop {
        // Read batch of wikidata_ids from consolidate
        let mut batch: Vec<String> = Vec::with_capacity(BATCH_SIZE);
        {
            let mut stmt = conn.prepare_cached(
                "SELECT wikidata_id FROM consolidate ORDER BY rowid LIMIT ?1 OFFSET ?2",
            )?;
            let rows = stmt.query_map(params![BATCH_SIZE as i64, offset], |r| {
                r.get::<_, String>(0)
            })?;
            for r in rows {
                batch.push(r?);
            }
        }

        if batch.is_empty() {
            break;
        }

        // For each wikidata_id, look up occupations_ids and compute flags
        let mut updates: Vec<(String, i32, i32)> = Vec::new();
        for wid in &batch {
            let occ_ids: Option<String> = {
                let mut stmt = conn.prepare_cached(
                    "SELECT occupations_ids FROM individuals_keys WHERE wikidata_id = ?1",
                )?;
                stmt.query_row(params![wid], |r| r.get::<_, Option<String>>(0))
                    .unwrap_or(None)
            };

            if let Some(occ_ids_str) = occ_ids {
                let mut is_sci = 0i32;
                let mut is_art = 0i32;
                for occ_id in occ_ids_str.split(';') {
                    let occ_id = occ_id.trim();
                    if let Some(meta) = occ_lookup.get(occ_id) {
                        match meta.as_str() {
                            "scientist" => is_sci = 1,
                            "artist" => is_art = 1,
                            _ => {}
                        }
                    }
                    if is_sci == 1 && is_art == 1 {
                        break;
                    }
                }
                if is_sci == 1 || is_art == 1 {
                    updates.push((wid.clone(), is_sci, is_art));
                    if is_sci == 1 {
                        scientist_count += 1;
                    }
                    if is_art == 1 {
                        artist_count += 1;
                    }
                }
            }
        }

        // Batch update
        if !updates.is_empty() {
            conn.execute_batch("BEGIN TRANSACTION;")?;
            {
                let mut stmt = conn.prepare_cached(
                    "UPDATE consolidate SET is_scientist = ?1, is_artist = ?2 WHERE wikidata_id = ?3",
                )?;
                for (wid, is_sci, is_art) in &updates {
                    stmt.execute(params![is_sci, is_art, wid])?;
                    updated_count += 1;
                }
            }
            conn.execute_batch("COMMIT;")?;
        }

        pb.inc(batch.len() as u64);
        offset += batch.len() as i64;

        if offset % 500_000 < BATCH_SIZE as i64 {
            log(&format!(
                "[45] Progress: {}/{} processed, {} updated (scientist={}, artist={})",
                offset, total, updated_count, scientist_count, artist_count
            ));
        }
    }
    pb.finish();
    log(&format!(
        "[45] Phase 3 complete: {} updated, {} scientists, {} artists ({})",
        updated_count,
        scientist_count,
        artist_count,
        elapsed(total_start)
    ));

    // ========================================================
    // PHASE 4: Create indexes
    // ========================================================
    log("[45] Phase 4: Creating indexes...");
    let idx_start = Instant::now();
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_consolidate_polity ON consolidate(polity_name);",
    )?;
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_consolidate_year ON consolidate(impact_year);",
    )?;
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_consolidate_scientist ON consolidate(is_scientist);",
    )?;
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_consolidate_artist ON consolidate(is_artist);",
    )?;
    log(&format!(
        "[45] Indexes created ({})",
        elapsed(idx_start)
    ));

    // ========================================================
    // PHASE 5: Export to CSV
    // ========================================================
    log("[45] Phase 5: Exporting to CSV...");
    let csv_start = Instant::now();

    let final_count: i64 =
        conn.query_row("SELECT COUNT(*) FROM consolidate", [], |r| r.get(0))?;

    let mut csv_file = fs::File::create(CSV_PATH)?;
    // Write header
    writeln!(
        csv_file,
        "wikidata_id,name_en,impact_year,polity_name,occupations,gender,references_count,is_scientist,is_artist"
    )?;

    let pb2 = ProgressBar::new(final_count as u64);
    pb2.set_style(
        ProgressStyle::default_bar()
            .template("{msg} [{bar:40}] {pos}/{len} ({eta})")
            .unwrap(),
    );
    pb2.set_message("Exporting CSV");

    let mut csv_offset: i64 = 0;
    let mut csv_rows_written = 0u64;

    loop {
        let mut rows_data: Vec<(
            String,
            String,
            String,
            String,
            String,
            String,
            String,
            i32,
            i32,
        )> = Vec::with_capacity(BATCH_SIZE);
        {
            let mut stmt = conn.prepare_cached(
                "SELECT wikidata_id, name_en, impact_year, polity_name, occupations,
                        gender, references_count, is_scientist, is_artist
                 FROM consolidate
                 ORDER BY rowid
                 LIMIT ?1 OFFSET ?2",
            )?;
            let rows = stmt.query_map(params![BATCH_SIZE as i64, csv_offset], |r| {
                Ok((
                    r.get::<_, String>(0).unwrap_or_default(),
                    r.get::<_, Option<String>>(1)
                        .unwrap_or(None)
                        .unwrap_or_default(),
                    r.get::<_, Option<i32>>(2)
                        .unwrap_or(None)
                        .map_or(String::new(), |v| v.to_string()),
                    r.get::<_, Option<String>>(3)
                        .unwrap_or(None)
                        .unwrap_or_default(),
                    r.get::<_, Option<String>>(4)
                        .unwrap_or(None)
                        .unwrap_or_default(),
                    r.get::<_, Option<String>>(5)
                        .unwrap_or(None)
                        .unwrap_or_default(),
                    r.get::<_, Option<i32>>(6)
                        .unwrap_or(None)
                        .map_or(String::new(), |v| v.to_string()),
                    r.get::<_, i32>(7).unwrap_or(0),
                    r.get::<_, i32>(8).unwrap_or(0),
                ))
            })?;
            for r in rows {
                rows_data.push(r?);
            }
        }

        if rows_data.is_empty() {
            break;
        }

        for (wid, name, year, polity, occs, gender, refs, is_sci, is_art) in &rows_data {
            writeln!(
                csv_file,
                "{},{},{},{},{},{},{},{},{}",
                csv_escape(wid),
                csv_escape(name),
                year,
                csv_escape(polity),
                csv_escape(occs),
                csv_escape(gender),
                refs,
                is_sci,
                is_art,
            )?;
            csv_rows_written += 1;
        }

        pb2.inc(rows_data.len() as u64);
        csv_offset += rows_data.len() as i64;
    }
    pb2.finish();
    csv_file.flush()?;
    log(&format!(
        "[45] CSV exported to {} ({} rows, {})",
        CSV_PATH,
        csv_rows_written,
        elapsed(csv_start)
    ));

    // ========================================================
    // PHASE 6: Final statistics
    // ========================================================
    log("[45] === Final Statistics ===");
    log(&format!("[45] Total rows in consolidate: {}", final_count));
    log(&format!("[45] Scientists: {}", scientist_count));
    log(&format!("[45] Artists: {}", artist_count));

    // Both scientist and artist
    let both: i64 = conn.query_row(
        "SELECT COUNT(*) FROM consolidate WHERE is_scientist = 1 AND is_artist = 1",
        [],
        |r| r.get(0),
    )?;
    log(&format!("[45] Both scientist and artist: {}", both));

    // Neither
    let neither: i64 = conn.query_row(
        "SELECT COUNT(*) FROM consolidate WHERE is_scientist = 0 AND is_artist = 0",
        [],
        |r| r.get(0),
    )?;
    log(&format!(
        "[45] Neither scientist nor artist (writers/other): {}",
        neither
    ));

    // Top 10 polities
    {
        let mut stmt = conn.prepare(
            "SELECT polity_name, COUNT(*) as cnt FROM consolidate
             GROUP BY polity_name ORDER BY cnt DESC LIMIT 10",
        )?;
        let rows: Vec<(String, i64)> = stmt
            .query_map([], |r| {
                Ok((
                    r.get::<_, Option<String>>(0)?.unwrap_or_default(),
                    r.get(1)?,
                ))
            })?
            .filter_map(|r| r.ok())
            .collect();
        log("[45] Top 10 polities:");
        for (polity, cnt) in &rows {
            log(&format!("[45]   {} -> {}", polity, cnt));
        }
    }

    // Gender distribution
    {
        let mut stmt = conn.prepare(
            "SELECT gender, COUNT(*) as cnt FROM consolidate
             GROUP BY gender ORDER BY cnt DESC LIMIT 10",
        )?;
        let rows: Vec<(String, i64)> = stmt
            .query_map([], |r| {
                Ok((
                    r.get::<_, Option<String>>(0)?.unwrap_or_default(),
                    r.get(1)?,
                ))
            })?
            .filter_map(|r| r.ok())
            .collect();
        log("[45] Gender distribution:");
        for (gender, cnt) in &rows {
            log(&format!("[45]   {} -> {}", gender, cnt));
        }
    }

    // Sample rows
    {
        let mut stmt = conn.prepare(
            "SELECT wikidata_id, name_en, impact_year, polity_name, occupations, gender, references_count, is_scientist, is_artist
             FROM consolidate LIMIT 5",
        )?;
        let rows = stmt.query_map([], |r| {
            Ok((
                r.get::<_, String>(0)?,
                r.get::<_, Option<String>>(1)?,
                r.get::<_, Option<i32>>(2)?,
                r.get::<_, Option<String>>(3)?,
                r.get::<_, Option<String>>(4)?,
                r.get::<_, Option<String>>(5)?,
                r.get::<_, Option<i32>>(6)?,
                r.get::<_, i32>(7)?,
                r.get::<_, i32>(8)?,
            ))
        })?;
        log("[45] Sample rows:");
        for r in rows {
            let (wid, name, year, polity, occs, gender, refs, is_sci, is_art) = r?;
            log(&format!(
                "[45]   {} | {} | {} | {} | {} | {} | refs={} | sci={} art={}",
                wid,
                name.unwrap_or_default(),
                year.unwrap_or(0),
                polity.unwrap_or_default(),
                occs.as_deref().unwrap_or("").chars().take(40).collect::<String>(),
                gender.unwrap_or_default(),
                refs.unwrap_or(0),
                is_sci,
                is_art,
            ));
        }
    }

    log(&format!(
        "=== Step 45 complete ({}) ===",
        elapsed(total_start)
    ));
    Ok(())
}

File: scripts/database_integration_scripts/legacy/45_create_consolidate.py
"""45 - Build the consolidate table + CSV.

Mirrors `enhance_db/src/bin/45_create_consolidate.rs`.

  Inputs : individuals_cliopatria (wikidata_id, name_en, polity_name, impact_date),
           individuals (occupations_en, gender, identifiers_count),
           individuals_keys (occupations_ids), occupations (id, meta_occupation).
  Output : consolidate (wikidata_id PK, name_en, impact_year, polity_name,
           occupations, gender, references_count, is_scientist, is_artist)
           + 4 indexes; CSV at data/consolidate.csv (full mode) or in the
           temp dir (sample mode).

Usage
-----
    python3 45_create_consolidate.py            # synthetic
    python3 45_create_consolidate.py --full     # real DB + data/consolidate.csv
"""

from __future__ import annotations

import csv
import sqlite3
import tempfile
from pathlib import Path

from tqdm import tqdm

from common import DATA_DIR, insert_rows, log, open_db, parse_run_mode

BATCH_SIZE = 50_000


def run(conn: sqlite3.Connection, csv_path: Path) -> int:
    log("[DB] 45: Building consolidate...")

    occ_meta: dict[str, str] = {}
    for oid, meta in conn.execute(
        "SELECT id, meta_occupation FROM occupations WHERE meta_occupation IS NOT NULL"
    ):
        occ_meta[oid] = meta
    log(f"[45] occupation meta: {len(occ_meta)}")

    conn.execute("DROP TABLE IF EXISTS consolidate")
    conn.execute(
        """
        CREATE TABLE consolidate (
            wikidata_id TEXT PRIMARY KEY,
            name_en TEXT,
            impact_year INTEGER,
            polity_name TEXT,
            occupations TEXT,
            gender TEXT,
            references_count INTEGER,
            is_scientist INTEGER DEFAULT 0,
            is_artist INTEGER DEFAULT 0
        )
        """
    )
    inserted = conn.execute(
        """
        INSERT INTO consolidate
            (wikidata_id, name_en, impact_year, polity_name, occupations, gender, references_count)
        SELECT ic.wikidata_id, ic.name_en, ic.impact_date, ic.polity_name,
               i.occupations_en, i.gender, i.identifiers_count
        FROM individuals_cliopatria ic
        JOIN individuals i ON ic.wikidata_id = i.wikidata_id
        """
    ).rowcount
    conn.commit()
    log(f"[45] inserted {inserted}")

    sci_count = art_count = updated = 0
    rows = conn.execute(
        "SELECT c.wikidata_id, k.occupations_ids "
        "FROM consolidate c LEFT JOIN individuals_keys k ON c.wikidata_id = k.wikidata_id"
    ).fetchall()
    buf: list[tuple] = []
    for wid, occ_ids in tqdm(rows, desc="45_flags"):
        if not occ_ids:
            continue
        is_sci = 0
        is_art = 0
        for oid in occ_ids.split(";"):
            meta = occ_meta.get(oid.strip())
            if meta == "scientist":
                is_sci = 1
            elif meta == "artist":
                is_art = 1
            if is_sci and is_art:
                break
        if is_sci or is_art:
            buf.append((is_sci, is_art, wid))
            sci_count += is_sci
            art_count += is_art
            updated += 1
            if len(buf) >= BATCH_SIZE:
                conn.executemany(
                    "UPDATE consolidate SET is_scientist = ?, is_artist = ? WHERE wikidata_id = ?",
                    buf,
                )
                conn.commit()
                buf.clear()
    if buf:
        conn.executemany(
            "UPDATE consolidate SET is_scientist = ?, is_artist = ? WHERE wikidata_id = ?",
            buf,
        )
        conn.commit()
    log(f"[45] flagged sci={sci_count} art={art_count} (updated {updated})")

    for ddl in (
        "CREATE INDEX IF NOT EXISTS idx_consolidate_polity ON consolidate(polity_name)",
        "CREATE INDEX IF NOT EXISTS idx_consolidate_year ON consolidate(impact_year)",
        "CREATE INDEX IF NOT EXISTS idx_consolidate_scientist ON consolidate(is_scientist)",
        "CREATE INDEX IF NOT EXISTS idx_consolidate_artist ON consolidate(is_artist)",
    ):
        conn.execute(ddl)
    conn.commit()

    csv_path.parent.mkdir(parents=True, exist_ok=True)
    cur = conn.execute(
        "SELECT wikidata_id, name_en, impact_year, polity_name, occupations, "
        "gender, references_count, is_scientist, is_artist FROM consolidate ORDER BY rowid"
    )
    with open(csv_path, "w", newline="", encoding="utf-8") as fh:
        w = csv.writer(fh)
        w.writerow([
            "wikidata_id", "name_en", "impact_year", "polity_name",
            "occupations", "gender", "references_count", "is_scientist", "is_artist",
        ])
        n_csv = 0
        for row in tqdm(cur, total=inserted, desc="45_csv"):
            w.writerow(row)
            n_csv += 1
    log(f"[45] wrote {n_csv} rows to {csv_path}")
    return inserted


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        db = Path(tmp) / "sample.sqlite3"
        csv_path = Path(tmp) / "consolidate.csv"
        with sqlite3.connect(db) as seed:
            seed.executescript(
                """
                CREATE TABLE individuals_cliopatria (wikidata_id TEXT PRIMARY KEY,
                    name_en TEXT, polity_name TEXT, impact_date INTEGER);
                CREATE TABLE individuals (wikidata_id TEXT PRIMARY KEY, name_en TEXT,
                    occupations_en TEXT, gender TEXT, identifiers_count INTEGER);
                CREATE TABLE individuals_keys (wikidata_id TEXT PRIMARY KEY,
                    occupations_ids TEXT);
                CREATE TABLE occupations (id TEXT PRIMARY KEY, meta_occupation TEXT);
                """
            )
            insert_rows(seed, "individuals_cliopatria", [
                {"wikidata_id": "Q1", "name_en": "Alice", "polity_name": "France", "impact_date": 1850},
                {"wikidata_id": "Q2", "name_en": "Bob", "polity_name": "USA", "impact_date": 1900},
                {"wikidata_id": "Q3", "name_en": "Cleo", "polity_name": "Egypt", "impact_date": -50},
            ])
            insert_rows(seed, "individuals", [
                {"wikidata_id": "Q1", "name_en": "Alice", "occupations_en": "physicist",
                 "gender": "female", "identifiers_count": 5},
                {"wikidata_id": "Q2", "name_en": "Bob", "occupations_en": "painter",
                 "gender": "male", "identifiers_count": 12},
                {"wikidata_id": "Q3", "name_en": "Cleo", "occupations_en": "monarch",
                 "gender": "female", "identifiers_count": 99},
            ])
            insert_rows(seed, "individuals_keys", [
                {"wikidata_id": "Q1", "occupations_ids": "Q169470"},
                {"wikidata_id": "Q2", "occupations_ids": "Q1028181"},
                {"wikidata_id": "Q3", "occupations_ids": "Q116"},
            ])
            insert_rows(seed, "occupations", [
                {"id": "Q169470", "meta_occupation": "scientist"},
                {"id": "Q1028181", "meta_occupation": "artist"},
                {"id": "Q116", "meta_occupation": "ruler"},
            ])
        with open_db(db) as conn:
            n = run(conn, csv_path=csv_path)
            rows = conn.execute(
                "SELECT wikidata_id, polity_name, is_scientist, is_artist FROM consolidate"
            ).fetchall()
        log(f"[sample] {n} rows")
        for r in rows:
            log(f"  {r}")
        log(f"  csv at {csv_path} ({csv_path.stat().st_size} bytes)")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn, csv_path=DATA_DIR / "consolidate.csv")
    else:
        _sample_main()
